File: src/lubenets/vladyslav/java/union/find/QuickUnionUF.java
package lubenets.vladyslav.java.union.find;

/**
 * Created by vladla on 10/27/16.
 *
 * initialize - N
 * union - N
 * find - N
 *
 */
public class QuickUnionUF {

    private int[] id;

    public QuickUnionUF(int n) {
        id = new int[n];
        for (int i = 0; i < n; i++) {
            id[i] = i;
        }
    }

    private int root(int i) {
        while (i != id[i]) i = id[i];
        return i;
    }

    public boolean connected(int p, int q) {
        return root(p) == root(q);
    }

    public void union(int p, int q) {
        int i = root(p);
        int j = root(q);
        id[i] = j;
    }

}

File: src/lubenets/vladyslav/kotlin/union/find/QuickUnionUF.kt
package lubenets.vladyslav.kotlin.union.find

/**
 * Created by vladla on 10/27/16.
 *
 * initialize - N
 * union - N
 * find - N
 *
 */
class QuickUnionUF {
    private var id : IntArray

    constructor(n : Int) {
        id = IntArray(n)
        for (i in id) {
            id[i] = i
        }
    }

    fun root(i : Int) : Int {
        var k = i
        while (k != id[k]) {
            k = id[k]
        }
        return k
    }

    fun connected(p : Int, q : Int) : Boolean {
        return root(p) == root(q)
    }

    fun union(p : Int, q : Int) {
        val i = root(p)
        val j = root(q)
        id[i] = j
    }

}